### ISSUE TITLE
Trailer Location?

### ISSUE THREAD

**Issue by mikewesten** [CherryFloors/jellyfin-plugin-cinemamode#6]
I'm confused by your instructions which state that "The plugin will automatically find any trailers you have in your Jellyfin library"
which I take it to mean, all my movie items which have a working, local trailer to be found in their respective movie item directories will be found by this plugin?
But on the plugin's settings page it has this 

> "Trailer Pre-Rolls Folder"
> Specify location of trailer pre-roll videos. (Should contain only videso, no subdirectories)

So which is it? Does it "automatically find any trailers" in my jellyfin library or do they all need to be in one specified directory? Does it do both? Does the latter override the former when used?

It's not clear.

**Comment by CherryFloors** [comment]
In short to answer your question, "Trailer Pre-Rolls" are not "Trailers". They are a type of pre-roll meant to play before a block of trailers (Think "Now playing on Jellyfin..."). I rushed through the docs and in my mind I thought I had made this distinction clear at some point but turns out I did not. There is a description in the config page  for the `Enable trailer pre-roll` check box but it might be more helpful to spell this out at the top of the page.  I will make some updates to the config page and docs to make this more clear.

To recap, the plugin will play media in the following order:
1. Trailer Pre-Rolls: A single random video from this folder/channel will play  (Think "Now playing on Jellyfin...").
2. Trailers: A specified number of trailers you have stored with your movies will play (These are in the Jellyfin DB so the plugin can grab them)
3. Feature Pre-Rolls: A single random video from this folder/channel will play  (Think "Now tour feature presentation...").

**Comment by mikewesten** [comment]
Thanks for the clarification. For my purposes I'm mostly interested in getting a handful of random movies trailers to play. I have the trailers stored alongside their respective movie items in their respective directories. I haven't got it working yet, but looking forward to it.

**Comment by mikewesten** [comment]
Here's a log if you're interested.
Configuration options = Play 1 trailer. Nothing else is set.
Client is Linux Jellyfin Media Player on Jellyfin v.10.8.10
No trailer gets played back. Only the movie plays.

`[2023-07-21 02:16:33.333 -07:00] [ERR] Error processing request. URL "GET" "/Users/5077fb25c5ef42c58f904b63b3927e7c/Items/3c4289440694f8f1ba4513c7c5c64294/Intros".
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.SZArrayHelper.get_Item[T](Int32 index)
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at Jellyfin.Plugin.CinemaMode.IntroManager.GetRandomMovieTrailer(Boolean isplayed, User user)
   at Jellyfin.Plugin.CinemaMode.IntroManager.Get(BaseItem item, User user)+MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at Emby.Server.Implementations.Library.LibraryManager.<>c.<GetIntros>b__132_1(IEnumerable`1 i)
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.ToArray()
   at Jellyfin.Api.Controllers.UserLibraryController.GetIntros(Guid userId, Guid itemId)
   at lambda_method1097(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Jellyfin.Server.Middleware.ServerStartupMessageMiddleware.Invoke(HttpContext httpContext, IServerApplicationHost serverApplicationHost, ILocalizationManager localizationManager)
   at Jellyfin.Server.Middleware.WebSocketHandlerMiddleware.Invoke(HttpContext httpContext, IWebSocketManager webSocketManager)
   at Jellyfin.Server.Middleware.IpBasedAccessValidationMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager)
   at Jellyfin.Server.Middleware.LanFilteringMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager, IServerConfigurationManager serverConfigurationManager)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Jellyfin.Server.Middleware.QueryStringDecodingMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.ReDoc.ReDocMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Jellyfin.Server.Middleware.RobotsRedirectionMiddleware.Invoke(HttpContext httpContext)
   at Jellyfin.Server.Middleware.LegacyEmbyRouteRewriteMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)
   at Jellyfin.Server.Middleware.ResponseTimeMiddleware.Invoke(HttpContext context, IServerConfigurationManager serverConfigurationManager)
   at Jellyfin.Server.Middleware.ExceptionMiddleware.Invoke(HttpContext context)
[2023-07-21 02:16:33.347 -07:00] [INF] User policy for "mikewesten". EnablePlaybackRemuxing: True EnableVideoPlaybackTranscoding: True EnableAudioPlaybackTranscoding: True
[2023-07-21 02:16:33.349 -07:00] [INF] StreamBuilder.BuildVideoItem( Profile="Jellyfin Media Player", Path="Media/Movies/Midnight Ride (1990)/Midnight.Ride.1990.mp4", AudioStreamIndex=null, SubtitleStreamIndex=null ) => ( PlayMethod=DirectPlay, TranscodeReason=0 ) "media:/videos/3c428944-0694-f8f1-ba45-13c7c5c64294/stream.mov?MediaSourceId=3c4289440694f8f1ba4513c7c5c64294&Static=true&VideoCodec=h264&AudioCodec=ac3&AudioStreamIndex=2&api_key=<token>&Tag=c82b192e83ee824f89651b0e2e4a28f8"
[2023-07-21 02:16:33.504 -07:00] [INF] Adding playback tracker : SmVsbHlmaW5NZWRpYVBsYXllciAxLjYuMSAobGludXgteDg2XzY0IDIwLjMpfDE2NDcwMzQ4NDI3Nzk1-5077fb25c5ef42c58f904b63b3927e7c-3c4289440694f8f1ba4513c7c5c64294
[2023-07-21 02:16:33.505 -07:00] [INF] PlaybackTracker : Adding Start Event : 07/21/2023 02:16:33
[2023-07-21 02:16:33.505 -07:00] [INF] Creating StartPlaybackTimer Task
[2023-07-21 02:16:33.506 -07:00] [INF] StartPlaybackTimer : Entered
[2023-07-21 02:16:33.779 -07:00] [INF] Processing playback tracker : "SmVsbHlmaW5NZWRpYVBsYXllciAxLjYuMSAobGludXgteDg2XzY0IDIwLjMpfDE2NDcwMzQ4NDI3Nzk1-5077fb25c5ef42c58f904b63b3927e7c-3c4289440694f8f1ba4513c7c5c64294"`

**Comment by mikewesten** [comment]
Wondering if this is part of the issue. Pics related
![jellyfin-movie-trailers-not-found](https://github.com/CherryFloors/jellyfin-plugin-cinemamode/assets/73662866/dbd15a56-80fe-4aa4-813f-e9d0e9fd629a)
![Screenshot from 2023-07-21 04-11-56](https://github.com/CherryFloors/jellyfin-plugin-cinemamode/assets/73662866/ee5febe7-8c2f-4b96-a67e-867e7037e2a5)

Local trailers aren't accessible under the trailers tab in the movie media library, they're only accessible through the trailers button on the movie media item page, which plays your local trailer if you have one (or TMDB's remote youtube trailer if you don't)

**Comment by CherryFloors** [comment]
I took a look at the log which was helpful, thanks for posting. I cant say for sure I know whats happening because I was unable to reproduce this on my test server but I have a pretty good idea whats going wrong. 

The plugin querys the jellyfin db using a "HasTrailer" flag. It then picks a random movie from that list. That part of the selection protects against a list with zero entries. The plugin then takes the movie it picked from the list and gets its list of local trailers. Since there can be more than one local trailer for that movie, the plugin again picks a random entry. This selection does not protect against a list with length 0 and is the most likely cause of the exception. I was not anticipating this as a possibility. In all my testing the behavior I saw was that if a movie has trailers, it has a list of local trailers with length >= 1.

I tried a few different things on my test server to try and reproduce this like removing trailers but I was unable to do so. How are your local trailers named and where are they stored?

**Comment by CherryFloors** [comment]
The TMDB Trailer Channels plugin is not a requirement. The local trailer that you play through the movie media item page are the trailers that should play. I will note that some other plugins also invoke an intro provider and I am unsure how they will behave together. I was not able to reproduce your defect but I have added some additional code to protect against the false positives the Jellyfin DB was returning. Hopefully this fix will filter out the incorrect values allowing your trailers to playback. 

Because I was not able to replicate this behavior from my test server, I cant say for certain this will fix the issue. Its possible that all of your movies that have trailers also have a local trailers list with length equal to 0. If this is the case, trailers will still not playback and we would need to do some more investigating.

**Comment by CherryFloors** [comment]
I also changed the plugin's config page. Hopefully it is more clear now. Thanks for the feedback and beta testing.

**Comment by mikewesten** [comment]
That fixed it. Trailers are playing now. The enforced ratings code seems to be working as well. Nice job.
